### PR TITLE
[Core][Iga] Iga Modeler

### PR DESCRIPTION
**Pull request by tteschemacher**
Guys,

this is the lovely IgaModeler. Please see it as an example modeler of #6605 .

Any questions? Thx

### CODE REVIEW

**Comment by tteschemacher** [comment]
@philbucher do you have any idea why the windows compiler is failing?

**Comment by philbucher** [comment]
is it compiling locally?

**Comment by philbucher** [comment]
given the error I am guessing `KRATOS_API` is missing
~~~
2020-06-10T06:17:06.1561264Z KratosIgaApplication_CXX_unity_0_6.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl Kratos::IgaModeler::IgaModeler(void)" (__imp_??0IgaModeler@Kratos@@QEAA@XZ) referenced in function "public: __cdecl Kratos::KratosIgaApplication::KratosIgaApplication(void)" (??0KratosIgaApplication@Kratos@@QEAA@XZ) [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\KratosIgaApplication_unity.vcxproj]
2020-06-10T06:17:06.1607181Z KratosIgaApplication_CXX_unity_0_6.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: virtual __cdecl Kratos::IgaModeler::~IgaModeler(void)" (__imp_??1IgaModeler@Kratos@@UEAA@XZ) referenced in function "public: virtual __cdecl Kratos::KratosIgaApplication::~KratosIgaApplication(void)" (??1KratosIgaApplication@Kratos@@UEAA@XZ) [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\KratosIgaApplication_unity.vcxproj]
2020-06-10T06:17:06.1648654Z KratosIgaApplication_CXX_unity_0_6.obj : error LNK2001: unresolved external symbol "public: virtual class std::shared_ptr<class Kratos::Modeler> __cdecl Kratos::IgaModeler::Create(class Kratos::Model &,class Kratos::Parameters)const " (?Create@IgaModeler@Kratos@@UEBA?AV?$shared_ptr@VModeler@Kratos@@@std@@AEAVModel@2@VParameters@2@@Z) [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\KratosIgaApplication_unity.vcxproj]
2020-06-10T06:17:06.1693056Z KratosIgaApplication_CXX_unity_0_6.obj : error LNK2001: unresolved external symbol "public: virtual class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl Kratos::IgaModeler::Info(void)const " (?Info@IgaModeler@Kratos@@UEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\KratosIgaApplication_unity.vcxproj]
2020-06-10T06:17:06.1746022Z KratosIgaApplication_CXX_unity_0_6.obj : error LNK2001: unresolved external symbol "public: virtual void __cdecl Kratos::IgaModeler::PrintData(class std::basic_ostream<char,struct std::char_traits<char> > &)const " (?PrintData@IgaModeler@Kratos@@UEBAXAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@@Z) [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\KratosIgaApplication_unity.vcxproj]
2020-06-10T06:17:06.1802355Z KratosIgaApplication_CXX_unity_0_6.obj : error LNK2001: unresolved external symbol "public: virtual void __cdecl Kratos::IgaModeler::PrintInfo(class std::basic_ostream<char,struct std::char_traits<char> > &)const " (?PrintInfo@IgaModeler@Kratos@@UEBAXAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@@Z) [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\KratosIgaApplication_unity.vcxproj]
2020-06-10T06:17:06.1829582Z D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication\Custom\KratosIgaApplication.pyd : fatal error LNK1120: 6 unresolved externals [D:\a\Kratos\Kratos\build\Custom\applications\IgaApplication
~~~

**Comment by tteschemacher** [comment]
> 
> 
> is it compiling locally?

yes.

I have the `KRATOS_API` in the iga_modeler.h (`class KRATOS_API(KRATOS_CORE) IgaModeler : public Modeler`). Does it need to be added at more places?

**Comment by tteschemacher** [comment]
> 
> 
> > is it compiling locally?
> 
> yes.
> 
> I have the `KRATOS_API` in the iga_modeler.h (`class KRATOS_API(KRATOS_CORE) IgaModeler : public Modeler`). Does it need to be added at more places?

I see I should make it to the `KRATOS_IGA_APPLICATION`

**Comment by philbucher** [comment]
I am not sure if it will work like this, `IGA_APPLICATION` is not "defined"

I am not the expert but according to more up-to-date apps this is usually coming from CMake
also the IGA-App is not splitted into core/py-interface which might cause issues here
see e.g. this:
https://github.com/KratosMultiphysics/Kratos/blob/3133f20f942d78aa55b153579dba946fb566bc76/applications/StructuralMechanicsApplication/CMakeLists.txt#L54

=> I recommend doing what I did with the CoSimApp in #6788 in a separate PR, then it should work for sure
(this is also how the CMake looks if you create a new app)

**Comment by tteschemacher** [comment]
I just added only the recommended line of yours. With this it seemed to work already.

**Comment by philbucher** [comment]
> I just added only the recommended line of yours. With this it seemed to work already.

ok then for now I guess

Just to say that in the long run it is inevitable to do what I explained in my previous comment
(its not much work and can save some headaches ;) )

File: applications/IgaApplication/custom_modelers/iga_modeler.cpp
Comment (dagmawibekel) [inline applications/IgaApplication/custom_modelers/iga_modeler.cpp:32]: Does it make sense to also check if there is a physics file? Maybe doing something like it is done for the cad_model_part and analysis_model_part: 
```C++
KRATOS_ERROR_IF_NOT(mParameters.Has("physics_file_name"))
```

Comment (tteschemacher) [reply applications/IgaApplication/custom_modelers/iga_modeler.cpp:32]: Basically, that's what I am trying with this if call. When `"physics_file_name"` is not provided it considers it being `"physics.iga.json"`. I know that this is not fully bullet proof. But we'll need to gain a little bit of experience of how this will be called.

**Comment by tteschemacher** [comment]
This Modeler will be local in the `IgaApplication`. So, if you're fine with it we can merge it. We can modify things always as soon as we realize that we need modifications. 

I already have another bigger addition waiting (the strong boundary conditions).

**Review by philbucher — APPROVED** [review]
I would not block for longer

@tteschemacher maybe wait for approval of @dagmawibekel , up to you

**Comment by tteschemacher** [comment]
Thank you @philbucher, yes, even though that it is against my usual behavior ;), I will wait for @dagmawibekel .

File: applications/IgaApplication/custom_modelers/iga_modeler.cpp
Comment (dagmawibekel) [inline applications/IgaApplication/custom_modelers/iga_modeler.cpp:189]: I think the nameS with the trailing S can be unified. So that we just use "brep_id" or "brep_ids" independetly if there is just one or more. This can be done in the future.

**Review by dagmawibekel — APPROVED** [review]
Thanks for waiting. I am flattered

File: applications/IgaApplication/custom_modelers/iga_modeler.cpp
Comment (tteschemacher) [reply applications/IgaApplication/custom_modelers/iga_modeler.cpp:189]: yeah, I agree. Thanks for pointing it out.
But, I am also a fan of allowing as many options for typos as possible.

**Pull request merged**
